src/protocols/unreal2/protocol.rs

```diff
@@ -0,0 +1,308 @@
+        .into();
+
+        let mut start = 0;
+
+        // Check if it is a UCS-2 string
+        if length >= 0x80 {
+            ucs2 = true;
+
+            length = (length & 0x7f) * 2;
+
+            start += 1;
+
+            // For UCS-2 strings, some unreal 2 games randomly insert an extra 0x01 here,
+            // not included in the length. Skip it if present (hopefully this never happens
+            // legitimately)
+            if let Some(1) = data[start ..].first() {
+                start += 1;
+            }
+        }
+
+        // If UCS2 the first byte is the masked length of the string
+        let result = if ucs2 {
+            let string_data = &data[start .. start + length];
+            if string_data.len() != length {
+                return Err(PacketBad.context("Not enough data in buffer to read string"));
+            }
+
+            // When node decodes UCS2 it uses the UFT16LE encoding.
+            // https://github.com/nodejs/node/blob/2aaa21f9f684484edb54be30589c4af0b923cdef/lib/buffer.js#L637-L645
+            let (result, _, invalid_sequences) = UTF_16LE.decode(string_data);
+
+            if invalid_sequences {
+                return Err(PacketBad.context("UTF-8 string contained invalid character(s)"));
+            }
+
+            result
+        } else {
+            // Else the string is null-delimited latin1
+
+            // TODO: Replace this with delimiter finder helper
+            let position = data
+            // Create an iterator over the data.
+                .iter()
+                // Find the position of the delimiter
+                .position(|&b| b == delimiter.as_ref()[0])
+                // If the delimiter is not found, use the whole data slice.
+                .unwrap_or(data.len());
+
+            length = position + 1;
+
+            // Decode as latin1
+            let (result, _, invalid_sequences) = WINDOWS_1252.decode(&data[0 .. position]);
+
+            if invalid_sequences {
+                return Err(PacketBad.context("latin1 string contained invalid character(s)"));
+            }
+
+            result
+        };
+
+        // Strip color encodings
+        // TODO: Improve efficiency
+        // TODO: There might be a nicer way to do this once string patterns are stable
+        //       https://github.com/rust-lang/rust/issues/27721
+
+        // After '0x1b' skip 3 characters (including the '0x1b')
+        let mut char_skip = 0usize;
+        let result: String = result
+            .chars()
+            .filter(|c: &char| {
+                if '\x1b'.eq(c) {
+                    char_skip = 4;
+                    return false;
+                }
+                char_skip = char_skip.saturating_sub(1);
+
+                char_skip == 0
+            })
+            .collect();
+
+        // Remove all characters between 0x00 and 0x1a
+        let result = result.replace(|c: char| c > '\x00' && c <= '\x1a', "");
```

Comment:
I'm not sure this belongs in the string decoder, do you think I should move it to a separate helper?

Reply:
As I think this file would get refactored/modified multiple times, we could leave it here for now.